### Домашнее задание к лекции "Случайные величины"
http://prntscr.com/uh35yg

Смоделировать игру против лотерейного автомата типа “777”. Игрок платит 1 руб., после чего выпадает случайное целое число, равномерно распределенное от 0 до 999. При некоторых значениях числа игрок получает выигрыш (см. справа)

Выгодна ли игра игроку?

Сколько в среднем приобретает
или теряет игрок за одну игру?

- 777: 200 руб.
- 999: 100 руб.
- 555: 50 руб.
- 333: 15 руб.
- 111: 10 руб.
- *77: 5 руб.
- **7: 3 руб.
- *00: 2 руб.
- **0: 1 руб.

'*' любая цифра

Дополнительное задание повышенной сложности. Теоретически рассчитать средний выигрыш (проигрыш) и сравнить с результатами моделирования

In [51]:
import numpy as np
import pandas as pd

In [41]:
def check_prizes(g):
    if g == 777:
        return 200
    if g == 999:
        return 100
    if g == 555:
        return 50
    if g == 333:
        return 15
    if g == 111:
        return 10
    if (g-77) % 100 == 0:
        return 5
    if (g-7) % 10 == 0:
        return 3
    if g % 100 == 0:
        return 2
    if g % 10 == 0:
        return 1
    return 0

In [ ]:
#def game_bandit

In [63]:
#cash = 50
#loss,find = [0,0]
test = 100000
df = []

for i in range(test):
    cash = 50
    loss,find = [0,0]
    while cash > 0:
        cash -= 1    
        x = np.random.randint(0,1000)    
        cash += check_prizes(x)    
        #print(x,check_prizes(x),cash)
        loss -= 1
        find += check_prizes(x)

    #print('Game Over')
    #print(loss,find)
    df.append([loss,find])

i = pd.DataFrame(df,columns = ('Loss','Find'))
print(i.mean())

Loss   -248.9618
Find    198.9618
dtype: float64


1. Игра не выгодна игроку на 100тыс. игр при старте с 50 рублях  в среднем получаем 199 р. теряем 249 рублей. Т.е. рано или поздно игра заканчивается пройгрышем всех денег.
2. В среднем игрок теряет 249 рублей за игру при старте с 50 рублями. Приобретает 199 рублей.

То есть если возмем полную вероятность получим,что шанс:
* Выйграть - 20,5%
* Проиграть - 79,5%

Если у нас будет 1р и мы будет играть получается что мы будем выигрывать 20 из 80 раз...
Минимальная сумма выйгрыша 1р и 3р, с более высокой вероятностью соотвественно у нас шанс выйграть 
40 рублей(18,1%) и проиграть 80 рублей.
В любом варианте игра не выгодна.

Если расчитать шансы на выйгрышь:
0,1% выйграть 200,150,100, 50, рублей...

Соотвественно полная вероятность:
0,1%*200+0,1%*150+0,1%*100+0,1%*50+0.1%*10+0,9%*5+8,1%*1+1%*5+10%*3


In [65]:
1/999*100

0.10010010010010009

In [66]:
9/999*100

0.9009009009009009

In [67]:
81/999*100

8.108108108108109

In [68]:
10/999*100

1.001001001001001

In [74]:
c = 0
for i in range(1000):
    if (i-7) % 10 == 0:
        c += 1
print(c)

100


In [72]:
100/999*100

10.01001001001001

In [77]:
0.1+0.1+0.1+0.1+0.1+0.9+8.1+1+10

20.5

In [78]:
test = 100000
df = []

for i in range(test):
    cash = 1
    loss,find = [0,0]
    while cash > 0:
        cash -= 1    
        x = np.random.randint(0,1000)    
        cash += check_prizes(x)    
        #print(x,check_prizes(x),cash)
        loss -= 1
        find += check_prizes(x)

    #print('Game Over')
    #print(loss,find)
    df.append([loss,find])

i = pd.DataFrame(df,columns = ('Loss','Find'))
print(i.mean())

Loss   -5.06782
Find    4.06782
dtype: float64


In [79]:
i

Loss  Find
0        -1     0
1        -1     0
2        -4     3
3        -1     0
4        -4     3
...     ...   ...
99995    -1     0
99996   -10     9
99997    -1     0
99998    -1     0
99999    -1     0

[100000 rows x 2 columns]